### PR TITLE
cleanup duplicate python prechecks and add error codes for curl and gpg

### PR DESCRIPTION
**Pull request by NarineM**
@KrisBash , @Microsoft/omsagent-devs

### CODE REVIEW

File: installer/bundle/bundle_skel.sh [hunk 1]
```diff
@@ -805,6 +829,23 @@ if [ "$installMode" = "I" -o "$installMode" = "U" ]; then
...
+        if [ -z "${forceFlag}" ]; then
+            echo "Error: curl is not installed, installation cannot continue."
+            echo "You can run this shell bundle with --force; in this case, we will install omsagent," >&2
+            echo "but omsconfig (DSC configuration) will not be available and will need to be re-installed." >&2
+            cleanup_and_exit $INSTALL_CURL
+        else
+            echo "curl is not installed,, please install curl and re-install omsconfig (DSC configuration) later."
+            echo "Installation will continue without installing omsconfig."
+        fi
+	fi
```
Comment (lagalbra) [inline installer/bundle/bundle_skel.sh]: Indentation seems inconsistent here; please check that all indents are 4 spaces.

File: installer/bundle/bundle_skel.sh [hunk 2]
```diff
@@ -838,9 +879,8 @@ if [ -n "${checkVersionAndCleanUp}" ]; then
 
     # omsconfig
     versionInstalled=`getInstalledVersion omsconfig`
-    versionAvailable=`getVersionNumber $DSC_PKG omsconfig-`
-    if ! python_ctypes_installed > /dev/null 2>&1; then ctypes_text=" (No ctypes)"; fi
-    if shouldInstall_omsconfig; then shouldInstall="Yes"; else shouldInstall="No${ctypes_text}"; fi
+    versionAvailable=`getVersionNumber $DSC_PKG omsconfig-`    
+    if shouldInstall_omsconfig; then shouldInstall="Yes"; else shouldInstall="No (No ctypes)"; fi
```
Comment (lagalbra) [inline installer/bundle/bundle_skel.sh]: If `shouldInstall_omsconfig` fails, it's not necessarily because of ctypes. I think we should remove the reason text and just print "Yes" or "No"; otherwise we could make a helper function to check ctypes, curl, and gpg and return a reason text from that.

File: installer/bundle/bundle_skel.sh [hunk 3]
```diff
@@ -871,10 +911,8 @@ case "$installMode" in
 
             pkg_add_list $OMS_PKG omsagent
 
-            python_ctypes_installed
-            if [ $? -eq 0 ]; then
-                pkg_add_list $DSC_PKG omsconfig
-            fi
+            shouldInstall_omsconfig
```
Comment (lagalbra) [inline installer/bundle/bundle_skel.sh]: The update path uses the result of this line in `pkg_upd_list $DSC_PKG omsconfig $?`, but the `pkg_add_list` subroutine doesn't have a parameter like that. I believe we have to keep the check if the return type is 0 before we call `pkg_add_list`:

```
shouldInstall_omsconfig
if [ $? -eq 0 ]; then
    pkg_add_list $DSC_PKG omsconfig
fi
```

File: installer/bundle/bundle_skel.sh [hunk 4]
```diff
@@ -805,6 +829,23 @@ if [ "$installMode" = "I" -o "$installMode" = "U" ]; then
         echo "sed was not installed, installation cannot continue. Please install sed."
         cleanup_and_exit $INSTALL_SED
     fi
+    curl_installed
+    if [ $? -ne 0 ]; then
+        if [ -z "${forceFlag}" ]; then
+            echo "Error: curl is not installed, installation cannot continue."
```
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: It is actually not needed at all, echo command would not throw an error. 
It was probably copy paste for the python, but we never redirect the output from echo command

File: installer/bundle/bundle_skel.sh [hunk 5]
```diff
@@ -379,6 +381,24 @@ python_ctypes_installed() {
...
+curl_installed() {
+    # Check for curl (required for omsconfig)
+    preReqCurl=1
+    which curl 1> /dev/null 2> /dev/null
+    [ $? -eq 0 ] && preReqCurl=0
+
+    return $preReqCurl
+}
+
+gpg_installed() {
```
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: documented dsc requirements are out of scope for omsagent and unnecessary would require keeping these docs in sync. there is a separate repo for each of the dependencies.  So I would think that it should be doc'ed in corresponding repo.

File: installer/bundle/bundle_skel.sh [hunk 6]
```diff
@@ -61,6 +61,8 @@ UNSUPPORTED_OPENSSL=60
 INSTALL_PYTHON_CTYPES=61
 INSTALL_TAR=62
 INSTALL_SED=63
+INSTALL_CURL=64
+INSTALL_GPG=65
```
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: I am not sure why it is required to duplicate these docs in two different GitHub repos instead of having them linked? I will see if I can have that also fixed today. Most likely a separate PR since it belongs to extension repro.

File: installer/bundle/bundle_skel.sh (hunk 2 above)
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: I left the original verbose message as it was, but I agree we can just print "Yes" or "No" here.

File: installer/bundle/bundle_skel.sh (hunk 6 above)
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: Added new bundle error codes in omsagent repo in this PR. I am going to leave it to you to remove duplicated codes from extension repo since you originally created the table. Thanks.

**Comment by NarineM** [comment]
Thanks for the review and brining the azure doc to attention :)

File: installer/bundle/bundle_skel.sh [hunk 7]
```diff
@@ -369,8 +371,7 @@ compare_install_type()
 
 python_ctypes_installed() {
     # Check for Python ctypes library (required for omsconfig)
-    hasCtypes=1
-    echo "Checking for ctypes python module ..."
```
Comment (lagalbra) [inline installer/bundle/bundle_skel.sh]: Why did you remove this message? It's not a  critical thing to print; I'm just curious.

File: installer/bundle/bundle_skel.sh (hunk 4 above)
Comment (lagalbra) [inline installer/bundle/bundle_skel.sh]: The reason the python message has it is because `>&2` redirects to stderr. It's a small thing considering we print lots of warning/error messages to stdout instead. You've made all the lines consistent now, so it's not a big deal.

File: installer/bundle/bundle_skel.sh [hunk 8]
```diff
@@ -871,10 +910,7 @@ case "$installMode" in
 
             pkg_add_list $OMS_PKG omsagent
 
-            python_ctypes_installed
-            if [ $? -eq 0 ]; then
-                pkg_add_list $DSC_PKG omsconfig
-            fi
+            if shouldInstall_omsconfig; then pkg_add_list $DSC_PKG omsconfig; fi
```
Comment (lagalbra) [inline installer/bundle/bundle_skel.sh]: We should standardize the single-line if statement vs. mutli-line if block throughout the code; as it is, this looks fine.

**Comment by lagalbra** [comment]
Remember to rebase against master and squash commits before merging!
